lib/spack/spack/version.py

```diff
@@ -96,26 +90,82 @@ def coercing_method(a, b, *args, **kwargs):
     return coercing_method
 
 
+class VersionStrComponent(object):
+    # NOTE: this is intentionally not a UserString, the abc instanceof
+    #       check is slow enough to eliminate all gains
+    __slots__ = ['inf_ver', 'data']
+
+    def __init__(self, string):
+        self.inf_ver = None
+        self.data = string
+        if len(string) >= iv_min_len:
+            try:
+                self.inf_ver = infinity_versions.index(string)
+            except ValueError:
+                pass
+
+    def __hash__(self):
+        return hash(self.data)
+
+    def __str__(self):
+        return self.data
+
+    def __eq__(self, other):
+        if isinstance(other, VersionStrComponent):
+            return self.data == other.data
+        return self.data == other
+
+    def __lt__(self, other):
+        if isinstance(other, VersionStrComponent):
+            if self.inf_ver is not None:
+                if other.inf_ver is not None:
+                    return self.inf_ver > other.inf_ver
+                return False
+            if other.inf_ver is not None:
+                return True
+
+            return self.data < other.data
+
+        if self.inf_ver is not None:
+            return False
+
+        # Numbers are always "newer" than letters.
+        # This is for consistency with RPM.  See patch
+        # #60884 (and details) from bugzilla #50977 in
+        # the RPM project at rpm.org.  Or look at
+        # rpmvercmp.c if you want to see how this is
+        # implemented there.
+        if isinstance(other, int):
+            return True
+
+        # if we get here, other should be str
+        return self < VersionStrComponent(other)
+
+    def __gt__(self, other):
+        return not self.__lt__(other)
+
+
 class Version(object):
     """Class to represent versions"""
+    __slots__ = ['version', 'separators', 'string']
 
     def __init__(self, string):
         if not isinstance(string, str):
             string = str(string)
 
-        if not VALID_VERSION.match(string):
-            raise ValueError("Bad characters in version string: %s" % string)
-
         # preserve the original string, but trimmed.
         string = string.strip()
         self.string = string
 
-        # Split version into alphabetical and numeric segments
-        segments = SEGMENT_REGEX.findall(string)
-        self.version = tuple(int_if_int(seg) for seg in segments)
+        if not VALID_VERSION.match(string):
+            raise ValueError("Bad characters in version string: %s" % string)
```

Comment:
This new check (with the updated regex) will not validate input with e.g. spaces in the middle like:
```
foo 1.3.2
```
while I think before these were validated. I think this new behavior is to be considered correct, but can we add tests to make sure we define the semantics of that edge case?

Reply:
@tgamblin @becker33 See :point_up: to check if there are edge cases where this change in behavior may have an impact.

Reply:
That's true.  The old code only validated `foo` from that version though, so the majority of it didn't get validated at all.  For example, `1.2.foo (U!@*#T$*&!T^@#$YUIT!#$%&*Y^*&$%^!@*#^$I!@(*!@&$&$Y` was a valid version in the old code. Literally any string with a prefix in it that could possibly be part of a valid version was "valid", here's another favorite:

```
>>> print(re.match(r'[A-Za-z0-9_.-]+', '1\0\t\n\r!@#$%^&*()-_=+\|}{[]";:2'))
<re.Match object; span=(0, 1), match='1'>
```

The version number from that one parses as `1.2` in the old code.

If we want to support spaces or anything else as separators, I'm totally good reworking it, just want to be sure we validate the whole thing mainly.

Reply:
> If we want to support spaces or anything else as separators, I'm totally good reworking it, just want to be sure we validate the whole thing mainly.

@trws I think the PR is good as is. Can we add a test that ensures strings like:
```
1.2.foo (U!@*#T$*&!T^@#$YUIT!#$%&*Y^*&$%^!@*#^$I!@(*!@&$&$Y
foo 1.2.3
```
are not considered valid versions?